operations/configuration_management/Write_Ansible_Code_for_CFS.md

```diff
@@ -11,22 +11,55 @@ Help with Ansible can be found in the external Ansible documentation:
+
+### Reduce Wasted Time
+
+* Use `include_*` (dynamic re-use) to skip multiple tasks at once when using conditionals.  Ansible evaluates conditionals for every node in every task.  This includes when the conditional is applied to a block, or a role imported with `roles:` or the `import_role` task.  This is because these are static imports that are compiled at the beginning of the playbook, and the conditional is inherited by every task in the role or block.  Evaluating these conditionals for each task may only take a second or two, but across the hundreds of tasks that might be part of a playbook, this can add up to significant wasted time. Instead use dynamic imports with the `include_*` tasks.  Because these are evaluated at runtime, a conditional can skip the import of the role or tasks entirely, and is only evaluated once.  See the Ansible documentation on [Conditionals with re-use](https://docs.ansible.com/ansible/latest/user_guide/playbooks_conditionals.html#conditionals-with-re-use) and [Re-using files and roles](https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse.html#re-using-files-and-roles) for more information.
+* Avoid using the same CFS configuration/playbook for all nodes and relying on the `hosts` keyword to determine what tasks will run against each node.  Ansible will skip sections of the playbook that have a `hosts` target that doesn't match any nodes in the current inventory/limit, but when multiple types of nodes are configured at the same time with the same configuration, they may end up in the same batch and Ansible run.  This would mean that Ansible has to run through the sections for both types of nodes, taking more time than if the nodes were in separate batches and could skip past the unneeded code.
+
+### Other Tips
+
 * Use loops rather than individual tasks where modules are called multiple times. Some Ansible modules will optimize the command, such as grouping package installations into a single transaction \(Refer to the external [Ansible playbook loops](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html) documentation\).
+* Use Ansible retries for small, recoverable failures.  CFS supports retries on a large scale, but it takes far more time for CFS to detect a failed component and spin up a new session than it does for Ansible to retry a task.
+* Don't use Ansible retries for failures than take a long time to recover from.  Retrying for a significant amount of time on one node can hold up all the other successful nodes in a batch.  If you can't recover from a failure quickly, let the node fail and CFS will separate it out from the successful nodes when new sessions are started.
+* Avoid `any_errors_fatal`.  In addition to not working with all Ansible strategies, this can cause an Ansible run to exit early, and the nodes that didn't have the error will have to start from the beginning of the playbook in the next session.
```

Comment:
```suggestion
* Avoid `any_errors_fatal`. In addition to not working with all Ansible strategies, this can cause an Ansible run to exit early, and the nodes that did not have the error will have to start from the beginning of the playbook in the next session.
```